controllers/gce/firewalls/firewalls.go

```diff
@@ -92,15 +102,48 @@ func (fr *FirewallRules) Sync(nodePorts []int64, nodeNames []string) error {
 		return nil
 	}
 	glog.V(3).Infof("Firewall %v already exists, updating nodeports %v", name, nodePorts)
-	return fr.cloud.UpdateFirewall(firewall)
+	return fr.updateFirewall(firewall, ing)
+}
+
+func (fr *FirewallRules) createFirewall(f *compute.Firewall, ing *extensions.Ingress) error {
+	err := fr.cloud.CreateFirewall(f)
+	if utils.IsForbiddenError(err) && fr.cloud.OnXPN() {
+		gcloudCmd := gce.FirewallToGCloudCreateCmd(f, fr.cloud.NetworkProjectID())
+		glog.V(3).Infof("Could not create L7 firewall on XPN cluster. Raising event for cmd: %q", gcloudCmd)
+		fr.raiseFirewallChangeNeededEvent(ing, gcloudCmd)
+		return nil
+	}
+	return err
+}
+
+func (fr *FirewallRules) updateFirewall(f *compute.Firewall, ing *extensions.Ingress) error {
+	err := fr.cloud.UpdateFirewall(f)
+	if utils.IsForbiddenError(err) && fr.cloud.OnXPN() {
+		gcloudCmd := gce.FirewallToGCloudUpdateCmd(f, fr.cloud.NetworkProjectID())
+		glog.V(3).Infof("Could not update L7 firewall on XPN cluster. Raising event for cmd: %q", gcloudCmd)
+		fr.raiseFirewallChangeNeededEvent(ing, gcloudCmd)
+		return nil
+	}
+	return err
+}
+
+func (fr *FirewallRules) deleteFirewall(name string) error {
+	err := fr.cloud.DeleteFirewall(name)
+	if utils.IsForbiddenError(err) && fr.cloud.OnXPN() {
+		gcloudCmd := gce.FirewallToGCloudDeleteCmd(name, fr.cloud.NetworkProjectID())
+		glog.V(3).Infof("Could not delete L7 firewall on XPN cluster. %q needs to be ran.", gcloudCmd)
+		// An event cannot be raised because there's no ingress for reference.
```

Comment:
Another reason why we should not delete ingress resource from apiserver till all cloud resources are deleted: https://github.com/kubernetes/kubernetes/issues/32157

Reply:
Agreed